packages/lib-sourcify/src/Verification/Verification.ts

```diff
@@ -0,0 +1,691 @@
+            "It seems your contract's metadata hashes match but not the bytecodes. You should add all the files input to the compiler during compilation and remove all others. See the issue for more information: https://github.com/ethereum/sourcify/issues/618",
+            VerificationErrorCode.EXTRA_FILE_INPUT_BUG,
+          );
+        }
+      } catch (e: any) {
+        if (e instanceof VerificationError) {
+          throw e;
+        }
+        logWarn('Error checking for extra-file-input-bug', {
+          chain: this.sourcifyChain.chainId,
+          address: this.address,
+          error: e.message,
+        });
+        throw new VerificationError(
+          e.message,
+          VerificationErrorCode.FAILED_TO_CHECK_EXTRA_FILE_INPUT_BUG,
+        );
+      }
+    }
+
+    // Try to match onchain creation bytecode with compiled creation bytecode
+    if (this.creatorTxHash) {
+      try {
+        logDebug('Matching with creation tx', {
+          chain: this.sourcifyChain.chainId,
+          address: this.address,
+          creatorTxHash: this.creatorTxHash,
+        });
+
+        // Get creation transaction data
+        const creatorTx = await this.sourcifyChain.getTx(this.creatorTxHash);
+        this.blockNumber = creatorTx.blockNumber || undefined;
+        this.deployer = creatorTx.from;
+
+        const { creationBytecode, txReceipt } =
+          await this.sourcifyChain.getContractCreationBytecodeAndReceipt(
+            this.address,
+            this.creatorTxHash,
+            creatorTx,
+          );
+        this._onchainCreationBytecode = creationBytecode;
+        this.txIndex = txReceipt.index;
+
+        await this.matchWithCreationTx();
+      } catch (e: any) {
+        logWarn('Error matching with creation tx', {
+          chain: this.sourcifyChain.chainId,
+          address: this.address,
+          creatorTxHash: this.creatorTxHash,
+          error: e.message,
+        });
+      }
+    }
+
+    if (this.creationMatch !== null || this.runtimeMatch !== null) {
+      logInfo('Verified contract', {
+        address: this.address,
+        chainId: this.sourcifyChain.chainId,
+        runtimeMatch: this.runtimeMatch,
+        creationMatch: this.creationMatch,
+      });
+      return;
+    }
+
+    throw new VerificationError(
+      "The deployed and recompiled bytecode don't match.",
+      VerificationErrorCode.NO_MATCH,
+    );
+  }
+
+  private async matchBytecodes(
+    context: BytecodeMatchingContext,
+  ): Promise<BytecodeMatchingResult> {
+    // Here we use bytecodes from the context because they are already processed
+    let normalizedRecompiledBytecode = context.normalizedRecompiledBytecode;
+    const onchainBytecode = context.isCreation
+      ? this.onchainCreationBytecode
+      : this.onchainRuntimeBytecode;
+
+    const cborAuxdata = context.isCreation
+      ? this.compilation.creationBytecodeCborAuxdata
+      : this.compilation.runtimeBytecodeCborAuxdata;
+
```

Comment:
One optional thing that you could do if you like it. If we pass `onchainBytecode`, `cborAuxdata` and `linkReferences` in the context (transformations should not be needed anymore when you address my other comment), we can also make this function pure. Then, I would also add a `matchingFunction` to the context like this: 

- In matchWithRuntimeBytecode: `matchingFunction: (onchainBytecode, normalizedRecompiledBytecode) => onchainBytecode === normalizedRecompiledBytecode`
- In matchWithCreationTx: `matchingFunction: (onchainBytecode, normalizedRecompiledBytecode) => onchainBytecode.startsWith(normalizedRecompiledBytecode)`

Reply:
I prefer having matchBytecodes inside `Verification`, from my pov it's a grouping of multiple transformations functions otherwised called directly in `matchWithRuntimeBytecode` and `matchWithCreationTx`.